Added support for recursively discovering output datasets

nuwang opened the pull request:
This PR adds a recurse flag for discovering output datasets, so that directory trees containing output can be discovered.

Usage:
```
<discover_datasets pattern="__name__" directory="subdir1" ext="txt" recurse="true" />
```
ping @jmchilton

nuwang committed:
Added support for recursively discovering output datasets

nuwang committed:
Added schema entry for discover_datasets recurse attribute

mvdbeek commented on lib/galaxy/tools/parameters/output_collect.py:
```diff
@@ -480,37 +480,48 @@ def discover_files(output_name, tool_provided_metadata, extra_file_collectors, j
-    if extra_file_collector.directory:
-        directory = os.path.join(directory, extra_file_collector.directory)
+def discover_target_directory(dir_name, job_working_directory):
+    if dir_name:
```
This isn't equivalent to the previous version, and I guess this is why collecting outputs from tool provided metadata fails (see https://jenkins.galaxyproject.org/job/docker-framework/9815/).
In line 477 this function is called. I suppose changing the call in 477 to pass in `extra_file_collector.directory` instead of `extra_file_collector` may fix the failing tests.

nuwang replied:
@mvdbeek Thanks for catching that - I'd missed your comment and I've just pushed in the fix as suggested. I changed it around a few times and must have forgotten to update the parameter.

martenson replied:
it seems to me this method may still have problems xref: https://github.com/galaxyproject/galaxy/issues/5560

nuwang committed:
Fix discover_target_directory and pass correct parameter

Fix according to review comment: https://github.com/galaxyproject/galaxy/pull/5240/files#r159138549

nuwang commented:
@mvdbeek Is there a way to retrigger the build? The other [PR](https://github.com/galaxyproject/galaxy/pull/5242) passes fine after the latest changes, and it includes all then changes in this PR, so it looks like the latest commit got caught to a temporary Travis issue.

mvdbeek commented:
Closing and opening again will trigger the tests (at least on travis).

nuwang commented:
@mvdbeek Good to know, thanks. It's passing now.

mvdbeek reviewed (approved):
Thanks @nuwang, that's definitely a useful feature!

nuwang commented:
@mvdbeek Thanks very much for reviewing!
@jmchilton Is it too late to merge this into the 18.01 release? Ted was pretty keen to get the multi-file upload going so it would be great if we can still merge this in if we are not past the window.

martenson commented:
@nuwang this will make it to the 18.01, thank you for the contribution!
